Add metrics section to benchmark (#4972)
opensearch-project/documentation-website: _benchmark/metrics/index.md

---
layout: default
title: Metrics
nav_order: 25
has_children: true
---

# Metrics

After a workload completes, OpenSearch Benchmark stores all metric records within its metrics store. These metrics can be kept in memory or inside of an OpenSearch cluster. 

## Storing metrics

You set where metrics are stored using the [`datastore.type`](https://opensearch.org/docs/latest/benchmark/configuring-benchmark/#results_publishing) parameter in your `benchmark.ini` file. 

### In memory

If you want to store your result in memory, use the following settings in the `results_publishing` section of `benchmark.ini`:

```json
[results_publishing]
datastore.type = in-memory
datastore.host = 
datastore.port = 
datastore.secure = False
datastore.ssl.verification_mode = 
datastore.user = 
datastore.password =
```

### OpenSearch

If you want to store your results in an external OpenSearch memory store, use the following settings in the `results_publishing` section of `benchmark.ini`:

> ```

Reviewer: `datastore.number_of_replicas` and `datastore.number_of_shards` settings are missing. Please add them. Also need to mentioned that if not provided then the defaults provided by OpenSearch are used which is 0 for replica and 1 for shard, please confirm. 
Also if the setting is applied after OpenSearch datastore cluster has been creaetd and results have been published, the replica and shards setting will only apply when new results indices are created at the start of each month.

[results_publishing]
datastore.type = opensearch
datastore.host = <opensearch endpoint>
datastore.port = 443
datastore.secure = True
datastore.ssl.verification_mode = none
datastore.user = <opensearch basic auth username>

> datastore.password = <opensearch basic auth password>

Author: ```suggestion
datastore.password = <opensearch basic auth password>
datastore.number_of_replicas = 1
datastore.number_of_shards = 2
```

```

> 

Author: ```suggestion
When `datastore.number_of_replicas` and `datastore.number_of_shards` are not provided, the defaults provided by OpenSearch are used, `0` for the number of replicas, and `1` for the number of shards. If these settings are changed after the datastore cluster has been created, the new replica and shards settings will only apply when new results indices are created at the end of the month. 

```

> After you run OpenSearch Benchmark configured to use OpenSearch as a datastore, OpenSeach Benchmark creates a `benchmark-metrics` Dashboard, where you can find your metric results by , as shown in the following image.

Reviewer: The OSB doesn't create any dashboard. OSB creates three indices, `benchmark-metrics-YYYY-MM` that holds granular metrics data and telemetry data, `benchmark-results-YYYY-MM` that holds the final results data and `benchmark-test-executions-YYYY-MM` that hold the data around execution-ids. 
The user can use OpenSearch Dashboards to visualize data on these indices.

> After you run OpenSearch Benchmark configured to use OpenSearch as a datastore, OpenSeach Benchmark creates a `benchmark-metrics` Dashboard, where you can find your metric results by , as shown in the following image.

Author: ```suggestion
After you run OpenSearch Benchmark configured to use OpenSearch as a datastore, OpenSeach Benchmark creates three indices:

- `benchmark-metrics-YYYY-MM`: Holds granular metrics and telemetry data.
- `benchmark-results-YYYY-MM`: Holds data based on final results.
```

![Dashboard index pattern]({{site.url}}{{site.baseurl}}/images/benchmark/metric-index-pattern.png).

## Next steps

> - For more information about how to design a metrics store, see [Metric records]({{site.url}}{{site.baseurl}}/benchmark/metrics/metric-records).

Author: ```suggestion
- For more information about how to design a metrics store, see [Metric records]({{site.url}}{{site.baseurl}}/benchmark/metrics/metric-records/).
```

> - For more information about what metrics are stored, see [Metric keys]({{site.url}}{{site.baseurl}}/benchmark/metrics/metric-keys).

Author: ```suggestion
- For more information about what metrics are stored, see [Metric keys]({{site.url}}{{site.baseurl}}/benchmark/metrics/metric-keys/).
```

---
layout: default
title: Metrics
nav_order: 25
has_children: true
---

# Metrics

After a workload completes, OpenSearch Benchmark stores all metric records within its metrics store. These metrics can be kept in memory or in an OpenSearch cluster. 

## Storing metrics

You can specify whether metrics are stored in memory or in a metrics store while running the benchmark by setting the [`datastore.type`](https://opensearch.org/docs/latest/benchmark/configuring-benchmark/#results_publishing) parameter in your `benchmark.ini` file. 

### In memory

If you want to store metrics in memory while running the benchmark, provide the following settings in the `results_publishing` section of `benchmark.ini`:

```ini
[results_publishing]
datastore.type = in-memory

> datastore.host = 

Author: ```suggestion
datastore.host = <host-url>
```

> datastore.port = 

Author: ```suggestion
datastore.port = <host-port>
```

datastore.secure = False

> datastore.ssl.verification_mode = 

Author: ```suggestion
datastore.ssl.verification_mode = <ssl-verification-details>
```

> datastore.user = 

Author: ```suggestion
datastore.user = <username>
```

> datastore.password =

Author: ```suggestion
datastore.password = <password>
```

```

### OpenSearch

If you want to store your results in an external OpenSearch memory store, use the following settings in the `results_publishing` section of `benchmark.ini`:

```ini
[results_publishing]
datastore.type = opensearch
datastore.host = <opensearch endpoint>
datastore.port = 443

> datastore.secure = True

Author: ```suggestion
datastore.secure = true
```

datastore.ssl.verification_mode = none
datastore.user = <opensearch basic auth username>
datastore.password = <opensearch basic auth password>
datastore.number_of_replicas = 
datastore.number_of_shards = 
```

> When `datastore.number_of_replicas` and `datastore.number_of_shards` are not provided, the defaults provided by OpenSearch are used, `0` for the number of replicas, and `1` for the number of shards. If these settings are changed after the datastore cluster has been created, the new replica and shards settings will only apply when new results indices are created at the end of the month. 

Author: We have to use "indices" in the case of OpenSearch Benchmark.

> 

Author: The metrics store and data stores are two separate concepts. Metrics store equals where OSB stores the resulting metrics. Data stores are where the indexed data used for Benchmarking lives.

After you run OpenSearch Benchmark configured to use OpenSearch as a data store, OpenSearch Benchmark creates three indexes:

> - `benchmark-metrics-YYYY-MM`: Holds granular metrics and telemetry data.

Author: ```suggestion
- `benchmark-metrics-YYYY-MM`: Holds granular metric and telemetry data.
```

- `benchmark-results-YYYY-MM`: Holds data based on final results.
- `benchmark-test-executions-YYYY-MM`: Holds data about `execution-ids`.

> You can visualize data inside these indices inside OpenSearch Dashboards.

Author: See my previous comment about indexes vs indices in OpenSearch Benchmark.

> You can visualize data inside these indices inside OpenSearch Dashboards.

Author: Lets go with in.
